internal/mkcgo/mkcgo_test.go

```diff
@@ -0,0 +1,402 @@
+				TypeDefs: []*mkcgo.TypeDef{
+					{"TD0", "void*"},
+					{"TD1", "const char*"},
+					{"TD2", "unsigned int*"},
+				},
+			},
+		}, {
+			content: `
+enum {
+	E0 = 1,
+	E1 = (1+1),
+	E2 =(1+2),
+};`,
+			want: &mkcgo.Source{
+				Enums: []*mkcgo.Enum{
+					{"E0", "1"},
+					{"E1", "(1+1)"},
+					{"E2", "(1+2)"},
+				},
+			},
+		}, {
+			content: `
+void F0(void) __attribute__((tag("t0"),noerror,tag("t1","tn0")));
+int F1(int p1) __attribute__((errcond("ec0"),  noescape,    nocallback));
+int * F2(int **p1, void  * p2);
+int *F3(int p1, void***) __attribute__((optional));
+unsigned   long F4(int func, int type, ...);
+int* F5(float, double) __attribute__((variadic("F4")));
+void F6();`,
+			want: &mkcgo.Source{
+				Funcs: []*mkcgo.Func{
+					{Name: "F0", Ret: "void", Params: []*mkcgo.Param{{"void", ""}}, FuncAttrs: mkcgo.FuncAttrs{Tags: []mkcgo.TagAttr{{Tag: "t0"}, {Tag: "t1", Name: "tn0"}}, NoError: true}},
+					{Name: "F1", Ret: "int", Params: []*mkcgo.Param{{"int", "p1"}}, FuncAttrs: mkcgo.FuncAttrs{ErrCond: "ec0", NoEscape: true, NoCallback: true}},
+					{Name: "F2", Ret: "int*", Params: []*mkcgo.Param{{"int**", "p1"}, {"void*", "p2"}}},
+					{Name: "F3", Ret: "int*", Params: []*mkcgo.Param{{"int", "p1"}, {"void***", ""}}, FuncAttrs: mkcgo.FuncAttrs{Optional: true}},
+					{Name: "F4", Ret: "unsigned long", Params: []*mkcgo.Param{{"int", "__func"}, {"int", "__type"}, {"...", ""}}},
+					{Name: "F5", Ret: "int*", Params: []*mkcgo.Param{{"float", ""}, {"double", ""}}, FuncAttrs: mkcgo.FuncAttrs{VariadicTarget: "F4"}},
+					{Name: "F6", Ret: "void", Params: []*mkcgo.Param{{"void", ""}}},
+				},
+			},
+		},
+	}
+	for i, tt := range tests {
+		// No need to specify a test name, the error message is enough to identify the test.
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			var got mkcgo.Source
+			if err := got.Parse(strings.NewReader(tt.content)); err != nil {
+				t.Fatal(err)
+			}
+			testSlice(t, "Includes", got.Includes, tt.want.Includes)
+			testSlice(t, "Comments", got.Comments, tt.want.Comments)
+			testSlice(t, "TypeDefs", got.TypeDefs, tt.want.TypeDefs)
+			testSlice(t, "Enums", got.Enums, tt.want.Enums)
+			testSlice(t, "Funcs", got.Funcs, tt.want.Funcs)
+		})
+	}
+}
+
+func TestParseError(t *testing.T) {
+	tests := []struct {
+		content string
+		want    string
+	}{
+		{
+			content: `/**/`,
+			want:    `block comments are not supported`,
+		}, {
+			content: "enum {\nE0 1,\n};",
+			want:    `can't parse enum in line "E0 1,": can't extract enum value`,
+		}, {
+			content: `typedef T0;`,
+			want:    `can't extract type name`,
+		}, {
+			content: `void foo(;`,
+			want:    `invalid signature`,
+		}, {
+			content: `void;`,
+			want:    `invalid signature`,
+		}, {
+			content: `void (void);`,
+			want:    `missing name`,
+		}, {
+			content: `void foo(,);`,
+			want:    `empty parameter`,
+		}, {
+			content: `void foo2(int a) __attribute__((variadic("foo")));`,
+			want:    "variadic target not added yet",
```

Comment:
This makes me realize "yet" can be misleading--it can imply that the parser knows that the target *will* be added later. But in this case, it won't ever be added.

```suggestion
			want:    "variadic target not found in preceding code",
```

(Or: could remove the order dependency by doing a second pass to link up the targets. Normally I'd say this is simply better, but we're in C land, so order dependency is something the user should be expecting. 😄)